homeassistant/components/ssdp/__init__.py

```diff
@@ -565,21 +570,30 @@ async def async_get_discovery_info_by_udn(self, udn: str) -> list[dict[str, Any]
 
 
 def discovery_info_from_headers_and_description(
-    info_with_desc: CaseInsensitiveDict,
-) -> dict[str, Any]:
+    combined_headers: Mapping[str, Any],
+    info_desc: Mapping[str, Any],
+) -> SsdpServiceInfo:
     """Convert headers and description to discovery_info."""
-    info = {
-        DISCOVERY_MAPPING.get(k.lower(), k): v
-        for k, v in info_with_desc.as_dict().items()
-    }
-
-    if ATTR_UPNP_UDN not in info and ATTR_SSDP_USN in info:
-        if udn := _udn_from_usn(info[ATTR_SSDP_USN]):
-            info[ATTR_UPNP_UDN] = udn
+    upnp_info = {**info_desc}
+    info = SsdpServiceInfo(
+        ssdp_usn=combined_headers["usn"],
+        ssdp_ext=combined_headers.get("ext"),
+        ssdp_server=combined_headers.get("server"),
+        ssdp_st=combined_headers.get("st"),
```

Comment:
If we decide that ssdp_st cannot be None, then I can adjust here to this, but it could generate a key error:
```suggestion
        ssdp_st=combined_headers.get("st", combined_headers["nt"]),
```

Reply:
@chishm has more experience with st vs nt messages.  Probably best to wait for his thoughts